qiskit/qasm3/exporter.py

```diff
@@ -760,21 +808,34 @@ def build_aliases(self, registers: Iterable[Register]) -> List[ast.AliasStatemen
             out.append(ast.AliasStatement(name, ast.IndexSet(elements)))
         return out
 
-    def build_quantum_instructions(self, instructions):
-        """Builds a list of call statements"""
-        ret = []
-        for instruction in instructions:
-            if isinstance(instruction.operation, ForLoopOp):
-                ret.append(self.build_for_loop(instruction))
-                continue
-            if isinstance(instruction.operation, WhileLoopOp):
-                ret.append(self.build_while_loop(instruction))
-                continue
-            if isinstance(instruction.operation, IfElseOp):
-                ret.append(self.build_if_statement(instruction))
-                continue
-            if isinstance(instruction.operation, SwitchCaseOp):
-                ret.extend(self.build_switch_statement(instruction))
+    def build_current_scope(self) -> list[ast.Statement]:
+        """Build the instructions that occur in the current scope.
+
+        In addition to everything literally in the circuit's ``data`` field, this also includes
+        declarations for any local :class:`.expr.Var` nodes.
+        """
+        scope = self.current_scope()
+
+        # We forward-declare all local variables; it's much easier to ensure that we don't introduce
+        # use-before-definition errors in the OQ3 output this way, if the user has side-stepped the
+        # `QuantumCircuit` API protection to produce a circuit that uses an uninitialised variable.
+        # (It would naturally be easier to see the def/use chain if we were using `DAGCircuit`.)
```

Comment:
This reads a bit awkwardly to me. Just to confirm the intent of forward declaring all the local variables is to protect against an error if the user manually generated a circuit which uses an uninitialized var? If so I wonder if just erroring is appropriate if that's not a valid program.

Reply:
The alternative to forwards-declare (`uint[8] x; x = 12;`) is to attempt to find the first use of the variable, and merge the declaration with its initialisation (`uint[8] x = 12;`).  It would be much more pleasant to do the latter, but there's like 200 subtleties in the model of `QuantumCircuit` to try and detect that without having a battle-tested def-use chain data structure / analysis that we can re-use from somewhere efficiently.  We do have that in `DAGCircuit`, but it would be too costly to pull the circuit into a DAG just to do that.

Edge-case examples: the variable might be declared in the outermost scope, but its first write is in both branches of an `if`/`else` - we still have to forward-declare.  Alternatively, there's an unintialised variable, so we have to have handling for them if we run off the end of the export and haven't put a declaration for them in.  I just kept coming up with more edge cases as I started writing this, so it's way easier just to forward-declare all the variables at the top and then not worry about it again.

Reply:
It's not an invalid program to have an uninitialised variable, it's just undefined behaviour in the spec.  For the control-flow scope thing as well - it's totally possible to have a variable that it guaranteed to be initialised before it's first read from, but has no single definition source point.

I clarified the comment in 3b4188f57b.